drivers/flash/flash_cadence_qspi_nor.c

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <device.h>
+
+#define DT_DRV_COMPAT cdns_qspi_nor
+
+#include <string.h>
+#include <drivers/flash.h>
+#include <init.h>
+#include <zephyr/logging/log.h>
+
+#include "flash_cadence_qspi_nor_ll.h"
+LOG_MODULE_REGISTER(flash_cadence, CONFIG_FLASH_LOG_LEVEL);
+
+struct flash_cad_priv {
+	DEVICE_MMIO_RAM;
+	struct cad_qspi_params params;
+};
+
+struct flash_cad_config {
+	DEVICE_MMIO_ROM;
+	uintptr_t cfg_data_reg;
+	uint32_t cfg_data_size;
+};
+
+static const struct flash_parameters flash_cad_parameters = {
+	.write_block_size = QSPI_BYTES_PER_DEV,
+	.erase_value = 0xff,
+};
+
+static int flash_cad_read(const struct device *dev, off_t offset,
+				void *data, size_t len)
+{
+	int rc;
+
+	if (dev == NULL) {
+		LOG_ERR("No device detected");
+		return -ENODEV;
+	}
+
+	struct flash_cad_priv *priv = dev->data;
+	struct cad_qspi_params *cad_params = &priv->params;
+
+	if ((data == NULL) || (len == 0)) {
+		LOG_ERR("Invalid input parameter for QSPI Write!");
+		return -EINVAL;
+	}
+
+	rc = cad_qspi_read(cad_params, data, offset, len);
+
+	if (rc < 0) {
+		LOG_ERR("Cadence QSPI Flash Read Failed");
+		return rc;
+	}
+
+	return 0;
+}
+
+static int flash_cad_erase(const struct device *dev, off_t offset,
+				size_t len)
+{
+	int rc;
+
+	if (dev == NULL) {
+		LOG_ERR("No device detected");
+		return -ENODEV;
+	}
+
+	struct flash_cad_priv *priv = dev->data;
+	struct cad_qspi_params *cad_params = &priv->params;
+
+	if (len == 0) {
+		LOG_ERR("Invalid input parameter for QSPI Write!");
+		return -EINVAL;
+	}
+
+	rc = cad_qspi_erase(cad_params, offset, len);
+
+	if (rc < 0) {
+		LOG_ERR("Cadence QSPI Flash Erase Failed!");
+		return rc;
+	}
+
+	return 0;
+}
+
+static int flash_cad_write(const struct device *dev, off_t offset,
+				const void *data, size_t len)
+{
+	int rc;
+
+	if (dev == NULL) {
+		LOG_ERR("No device detected!");
+		return -ENODEV;
+	}
+
+	struct flash_cad_priv *priv = dev->data;
+	struct cad_qspi_params *cad_params = &priv->params;
+
+	if ((data == NULL) || (len == 0)) {
+		LOG_ERR("Invalid input parameter for QSPI Write!");
+		return -EINVAL;
+	}
+
+	rc = cad_qspi_write(cad_params, (void *)data, offset, len);
+
+	if (rc < 0) {
+		LOG_ERR("Cadence QSPI Flash Write Failed!");
+		return rc;
+	}
+
+	return 0;
+}
+
+static const struct flash_parameters *
+flash_cad_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return &flash_cad_parameters;
+}
+
+static const struct flash_driver_api flash_cad_api = {
+	.erase = flash_cad_erase,
+	.write = flash_cad_write,
+	.read = flash_cad_read,
+	.get_parameters = flash_cad_get_parameters,
+};
+
+static int flash_cad_init(const struct device *dev)
+{
+
+	int rc;
+
+	if (dev == NULL) {
+		LOG_ERR("No device detected");
+		return -ENODEV;
+	}
```

Comment:
This should never happen right?

Reply:
This is the software development lifecycle requirement, we always need to check the incoming pointer to make sure it is not NULL to prevent the Kernel hang, just incase.